### PR TITLE
Fixes #406 memory leak in varies addEliminator calls

### LINKED ISSUE #406: FieldVerifier::eliminators has memory leak
```cpp
FieldVerifier::FieldVerifier() {
  this->eliminators.push_back(new HFRecordEliminator());
  this->eliminators.push_back(new HFSLookupEliminator());
  this->eliminators.push_back(new HFLookupEliminator());
  this->eliminators.push_back(new HFTEnvLookupEliminator());
}
```

Those memory are never got released. Those pointers cannot be simply replaced by smart pointers. Because some of pointers in `eliminator` might not be dynamically allocated. There is another API for adding,

```cpp
void FieldVerifier::addEliminator(HFEliminator* hfe) {
  this->eliminators.push_back(hfe);
  ...
}
```

it is called by

```cpp
ProcessP::ProcessO(FieldVerifier* fv) {
  fv->addEliminator(&this->procman);
}
```

and `procman` is __not__ dynamically allocated

```cpp
class ProcessP : public Unqualifier {
...
  ProcManager procman;
};
```

My first try was to use concrete member variables as member variables, and pass their pointers to `eliminator`. However, theirs corresponding headers have to be included in `FieldVerifier.H`, which causes recursive header include error.

To make things worse, `CtorVerifier`, `AppendsToUnqualifier` etc. are all using the same pattern.

The easiest (improper?) way I can think of is to keep current API and do it like this

```cpp
struct HFEliminator {
...
private:
   using HFEliminators = std::vector<HFEliminator*>;
   HFEliminators eliminators;

   std::shared_ptr<HFEliminator> hfRecordEliminator;
   std::shared_ptr<HFEliminator> hfsLookupEliminator;
   std::shared_ptr<HFEliminator> hfLookupEliminator;
   std::shared_ptr<HFEliminator> hftEnvLookupEliminator;
};
```

```cpp
FieldVerifier::FieldVerifier() {    
  hfRecordEliminator = std::make_shared<HFRecordEliminator>();    
  hfsLookupEliminator = std::make_shared<HFSLookupEliminator>();    
  hfLookupEliminator = std::make_shared<HFLookupEliminator>();    
  hftEnvLookupEliminator = std::make_shared<HFTEnvLookupEliminator>(); 

  this->eliminators.push_back(hfRecordEliminator.get());    
  this->eliminators.push_back(hfsLookupEliminator.get());    
  this->eliminators.push_back(hfLookupEliminator.get());    
  this->eliminators.push_back(hftEnvLookupEliminator.get());  
} 
```

It indeed fixes the memory leak (confirmed by valgrind), however, it feels hacky.

I'll try to figure out a better way and create a PR in the next few days.

### PR DESCRIPTION
**Pull request by mo-xiaoming**
And modernize code along the way

### CODE REVIEW

File: include/hobbes/eval/cc.H [hunk 1]
```diff
@@ -108,11 +95,13 @@ public:
   SearchEntries search(const std::string&, const MonoTypePtr&);
   SearchEntries search(const std::string&, const std::string&);
 private:
+  static std::recursive_mutex hccmtx;
```
Comment (smunix) [inline include/hobbes/eval/cc.H:98]: do _not_ use a statically initialized mutex

Comment (smunix) [reply include/hobbes/eval/cc.H:98]: You could define and use helpers such as these below:
```c++
static auto lock() -> std::recursive_mutex& {
  static std::recursive_mutex h;
  return h;
}
template<class Fn>
static auto hlock(Fn&& fn) -> decltype(fn()) {
  std::lock_guard<std::recursive_mutex> l{lock()};
  return fn();
}
```

File: include/hobbes/eval/cc.H [hunk 2]
```diff
@@ -172,20 +161,20 @@ public:
...
   // get the x86 machine code for an expression (useful for debugging)
-  typedef std::vector<unsigned char> bytes;
+  using bytes = std::vector<unsigned char>;
   bytes machineCodeForExpr(const std::string& expr);
 
   // keep track of C++ classes so that we can perform upcasts where necessary
   template <typename T>
     void addObj() {
-      hlock _;
+      std::lock_guard<std::recursive_mutex> lk{hccmtx};
```
Comment (smunix) [inline include/hobbes/eval/cc.H:170]: Then we can use the helpers previously defined to introduce a protected computation context:

```c++
template<typename T>
void addObj() {
  return hlock([this](){ return objs->add<T>(); })
}
```

File: include/hobbes/lang/preds/hasfield.H [hunk 3]
```diff
@@ -53,22 +54,22 @@ struct HFEliminator {
 class FieldVerifier : public Unqualifier {
 public:
   FieldVerifier();
-  static std::string constraintName();
+  static const std::string& constraintName();
```
Comment (smunix) [inline include/hobbes/lang/preds/hasfield.H:57]: what did trigger this change? how is this a memory leak issue?

File: include/hobbes/lang/preds/subtype/obj.H [hunk 4]
```diff
@@ -20,7 +21,7 @@ namespace hobbes {
 //   * the constant adjustment    (for A <: B, A : ..., B;    the memory for B is at an offset from A)
 //   * the vtbl lookup adjustment (for A <: B, A : virtual B; A holds a pointer to the base address for B)
 struct PtrAdjustment {
-  PtrAdjustment(bool vtblLookup, int offset, const std::string& targetTy) : vtblLookup(vtblLookup), offset(offset), targetTy(targetTy) { }
+  PtrAdjustment(bool vtblLookup, int offset, std::string targetTy) : vtblLookup(vtblLookup), offset(offset), targetTy(std::move(targetTy)) { }
```
Comment (smunix) [inline include/hobbes/lang/preds/subtype/obj.H:24]: what's the justification for this change?

File: include/hobbes/eval/cc.H [hunk 5]
```diff
@@ -287,20 +276,20 @@ private:
   SearchCache searchCache;
 
   // optimization options
-  bool runModInlinePass;
-  bool genInterpretedMatch;
-  bool checkMatchReachability;
-  bool lowerPrimMatchTables;
-  bool columnwiseMatches;
-  size_t maxExprDFASize={1000};
+  bool runModInlinePass       = true;
```
Comment (smunix) [inline include/hobbes/eval/cc.H:279]: how are these changes related to the eliminators memory leak issue?

File: include/hobbes/ipc/process.H [hunk 6]
```diff
@@ -11,21 +11,21 @@ namespace hobbes {
 // resolve process invocation constraints
 class ProcessP : public Unqualifier {
 public:
-  ProcessP(FieldVerifier*);
+  explicit ProcessP(FieldVerifier*);
 
-  static std::string constraintName();
+  static const std::string& constraintName();
```
Comment (smunix) [inline include/hobbes/ipc/process.H:16]: how is this related to the memory leak issue?

File: lib/hobbes/db/bindings.C [hunk 7]
```diff
@@ -119,9 +119,8 @@ unsigned int storedOffset(const MonoTypePtr& ty, const std::string& lbl) {
   FileConfig fcfg;
   if (unpackFileType(ty, &fcfg)) {
     return storedOffset(fcfg, lbl);
-  } else {
-    throw std::runtime_error("Internal error, can't determine stored offset of '" + lbl + "' in type " + show(ty));
   }
+  throw std::runtime_error("Internal error, can't determine stored offset of '" + lbl + "' in type " + show(ty));
```
Comment (smunix) [inline lib/hobbes/db/bindings.C:123]: ?

File: lib/hobbes/db/bindings.C [hunk 8]
```diff
@@ -944,8 +943,8 @@ static bool decLF(const ConstraintPtr& c, MonoTypePtr* lhs, MonoTypePtr* rhs) {
   return false;
 }
 
-#define READ_FILE_SYM "inputFile"
-#define WRITE_FILE_SYM "outputFile"
+static constexpr char const* READ_FILE_SYM = "inputFile";
```
Comment (smunix) [inline lib/hobbes/db/bindings.C:946]: ?

File: lib/hobbes/eval/cc.C [hunk 9]
```diff
@@ -24,37 +24,20 @@
...
 #include <fstream>
 #include <mutex>
 
 namespace hobbes {
 
-// protect access to hobbes/llvm resources
-static std::recursive_mutex hccmtx;
-hlock:: hlock() { hccmtx.lock();   }
-hlock::~hlock() { hccmtx.unlock(); }
+std::recursive_mutex cc::hccmtx{};
```
Comment (smunix) [inline lib/hobbes/eval/cc.C:35]: do not statically initialize

File: lib/hobbes/eval/cc.C [hunk 10]
```diff
@@ -24,37 +24,20 @@
...
-  unreachableMatchRowsPtr(nullptr),
-  runModInlinePass(true),
-  genInterpretedMatch(false),
-  checkMatchReachability(true),
-  lowerPrimMatchTables(false),
-  columnwiseMatches(false),
-  tenv(new TEnv()),
-  objs(new Objs())
-{
+cc::cc() {
```
Comment (smunix) [inline lib/hobbes/eval/cc.C:38]: why remove the initialization ?

File: lib/hobbes/lang/preds/hasctor.C [hunk 11]
```diff
@@ -5,33 +5,33 @@
...
 /////////////////////////////////////////////////////
 CtorVerifier::CtorVerifier() {
-  this->eliminators.push_back(new HCVariantEliminator());
+  addEliminator(std::make_shared<HCVariantEliminator>());
 }
 
-std::string CtorVerifier::constraintName() {
-  return "HasCtor";
+const std::string& CtorVerifier::constraintName() {
+  static const auto n = std::string("HasCtor");
```
Comment (smunix) [inline lib/hobbes/lang/preds/hasctor.C:22]: let's not take references over this name

File: lib/hobbes/lang/preds/hasctor.C [hunk 12]
```diff
@@ -5,33 +5,33 @@
...
 }
 
 HCEliminator* CtorVerifier::findEliminator(const TEnvPtr& tenv, const HasCtor& hc, Definitions* ds) const {
-  for (auto hce : this->eliminators) {
-    if (hce->satisfiable(tenv, hc, ds)) {
-      return hce;
-    }
-  }
-  return 0;
+  const auto it = std::find_if(eliminators.begin(), eliminators.end(), [&](auto& hce) {
```
Comment (smunix) [inline lib/hobbes/lang/preds/hasctor.C:31]: what's the justification for this change?

File: lib/hobbes/lang/preds/hasfield.C [hunk 13]
```diff
@@ -11,7 +11,7 @@
 
 namespace hobbes {
 
-#define FIELD_VALUE_FN "fieldValue"
+static constexpr char const* FIELD_VALUE_FN = "fieldValue";
```
Comment (smunix) [inline lib/hobbes/lang/preds/hasfield.C:14]: ?

**Review by smunix — CHANGES_REQUESTED** [review]

File: include/hobbes/eval/cc.H (hunk 1 above)
Comment (mo-xiaoming) [reply include/hobbes/eval/cc.H:98]: This helper idiom is much better than the original `hlock` design and using vanilla lock guard everywhere

Just one question, why _do not_ use statically initialized mutex? `h` in `lock()` is statically initialized as well, I don't know the difference

File: include/hobbes/lang/preds/hasfield.H (hunk 3 above)
Comment (mo-xiaoming) [reply include/hobbes/lang/preds/hasfield.H:57]: No, there is no memory leak.

I really should have separated this commit into two. One for the leak, another for modernization/optimization.

The original code
```cpp
std::string FieldVerifier::constraintName() {
  return "HasField";
}
```

creates a temporary `std::string` each time it returns, and this returned value is never going to be used to create new `std::string`, so RVO doesn't really help here. Although, small string optimization makes it cheap.

Although, if copy-on-write `std::string` is used on some platforms(rhel?), then this change probably makes no much difference

File: include/hobbes/lang/preds/subtype/obj.H (hunk 4 above)
Comment (mo-xiaoming) [reply include/hobbes/lang/preds/subtype/obj.H:24]: For the argument `targetTy`, the number is the number of constructor calls

| param type \ arg type| const char* | std::string (lvalue) | std::string (rvalue) |
|-|-|-|-|
|const std::string&| 2  | 1 | 1  |
|std::string + std::move | 1  | 1 | 0 |

Overall, this idiom gives reasonable better performance without adding extra constructor overloads or resort to templates

File: include/hobbes/eval/cc.H (hunk 5 above)
Comment (mo-xiaoming) [reply include/hobbes/eval/cc.H:279]: No, this is about modernization. Again, I really should have split this commit

`cc` already has enough member variables, and I just added another one in #408. init list in constructor before c++11 is easy to miss a member or get the order wrong.

That's why I think direct member initialization is a better choice in most cases, won't left member uninitialized and don't need to worry about initialization order

File: lib/hobbes/lang/preds/hasctor.C (hunk 12 above)
Comment (mo-xiaoming) [reply lib/hobbes/lang/preds/hasctor.C:31]: Just a minor refactoring, using standard algorithms is almost always easy to read and reason, IMO

File: lib/hobbes/lang/preds/hasfield.C (hunk 13 above)
Comment (mo-xiaoming) [reply lib/hobbes/lang/preds/hasfield.C:14]: There is no clear benefits for this change in this case. Just in general, I prefer C++ style instead of C

File: include/hobbes/eval/cc.H (hunk 1 above)
Comment (smunix) [reply include/hobbes/eval/cc.H:98]: Static initialization breaks linker dependency analysis, especially when it comes to lining against static libraries. This [blog post](https://www.bfilipek.com/2018/02/static-vars-static-lib.html) explains a little how tricky they are to get right.

File: include/hobbes/lang/preds/hasfield.H (hunk 3 above)
Comment (smunix) [reply include/hobbes/lang/preds/hasfield.H:57]: There's really no performance issue with this code, unless you've got benchmarks that show otherwise. 

While I appreciate the intent to squeeze all possible performance optimization out of any code, I do value more simplicity over complexity. RVO, SSO, COW, ..., all increase the entering barrier for Junior developers. Let's KISS this for now.

File: include/hobbes/eval/cc.H (hunk 5 above)
Comment (smunix) [reply include/hobbes/eval/cc.H:279]: Agreed, let's split.

Btw, aren't there any compiler flags to help catch members initialization order issues? Cam we have the compiler catch these instead?

Comment (mo-xiaoming) [reply include/hobbes/eval/cc.H:279]: I checked the git log of CMakeLists.txt, interestingly, `-Wreorder` was the first warning flag added explicitly. So, yes, as you pointed out, this kind of error will be caught.

However, still nothing prevents me from making this kind of mistake

```cpp
struct S {
    S() : b(0) {}
private:
    int a;
    int b;
};
```
[gotbolt](https://godbolt.org/z/qTaedqTGv)

File: include/hobbes/eval/cc.H (hunk 1 above)
Comment (mo-xiaoming) [reply include/hobbes/eval/cc.H:98]: You are right

First of all, I got the life time of a static class member wrong, I always thought it is like local static, turns out it is just like global static

```cpp
#include <cstdio>

struct X {
    X() { std::puts("X()"); }
};

struct S {
    static X x;
};

X S::x{};

int main() {}
```
[It prints `X()` without initialization `S`](https://godbolt.org/z/n4qonhWPr), I didn't know this

Secondly, as always, C++ surprises me on daily basis. Let's do it your way, the correct way :)

**Comment by mo-xiaoming** [comment]
strip my so-called 'modernize', actually PR unrelated code

**Review by smunix — APPROVED** [review]
